src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,371 @@
+    private Thread writeLockThread = null;
 
     /**
-     * This method removes an element from the list of "registered" elements,
-     * such that it will no longer be returned by calls to
-     * <tt>getRegisteredInstances</tt>.
-     *
-     * This method does not remove the element from the underlying queue; it
-     * is assumed that the caller has already done so via the methods of the
-     * parent class.
+     * Create a JRubyPool
      *
-     * @param e the element to remove from the list of registered instances.
+     * @param size maximum capacity for the pool.
      */
+    public JRubyPool(int size) {
+        liveQueue = new LinkedList<>();
+        maxSize = size;
+    }
+
     @Override
-    synchronized public void unregister(E e) throws InterruptedException {
-        registeredElements.remove(e);
+    public void register(E e) {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            if (registeredElements.size() == maxSize)
+                throw new IllegalStateException(
+                        "Unable to register additional instance, pool full");
+            registeredElements.add(e);
+            liveQueue.addLast(e);
+            signalNotEmpty();
+        } finally {
+            lock.unlock();
+        }
     }
 
+    @Override
+    public void unregister(E e) {
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            registeredElements.remove(e);
+            signalIfAllRegisteredInQueue();
+        } finally {
+            lock.unlock();
+        }
+    }
 
-    /**
-    * This method is analagous to <tt>takeFirst</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class, but also causes a read lock to be
-    * locked from the <tt>ReentrantReadWriteLock</tt>.
-    *
-    * @return The head of the deque.
-    *
-    * @throws IllegalStateException if the thread holding the write lock
-    *         attempts to call this method.
-    * @throws InterruptedException
-    */
     @Override
-    public E borrowItem() throws InterruptedException, IllegalStateException {
-        if (lock.isWriteLockedByCurrentThread()) {
-          throw new IllegalStateException("The current implementation has a risk of deadlock if you attempt to borrow a JRuby instance while holding the write lock!");
+    public E borrowItem() throws InterruptedException {
+        E item = null;
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            final Thread currentThread = Thread.currentThread();
+            do {
+                if (isWriteLockHeldByAnotherThread(currentThread)) {
+                    notWriteLocked.await();
+                } else if (liveQueue.size() < 1) {
+                    notEmpty.await();
+                } else {
+                    item = liveQueue.removeFirst();
+                }
+            } while (item == null);
```

Comment:
Doesn't make much difference to me, so I'd probably err on the side of leaving it alone so we can get this closed out sooner, since it's been open for a while.